lib/matplotlib/backends/qt_compat.py

```diff
@@ -1,220 +1,159 @@
 import sys
-from matplotlib import rcParams
 
-_log = logging.getLogger(__name__)
+from matplotlib import rcParams
 
-# Available APIs.
-QT_API_PYQT = 'PyQt4'       # API is not set here; Python 2.x default is V 1
-QT_API_PYQTv2 = 'PyQt4v2'   # forced to Version 2 API
-QT_API_PYSIDE = 'PySide'    # only supports Version 2 API
-QT_API_PYQT5 = 'PyQt5'      # use PyQt5 API; Version 2 with module shim
-QT_API_PYSIDE2 = 'PySide2'  # Version 2 API with module shim
-
-ETS = dict(pyqt=(QT_API_PYQTv2, 4), pyside=(QT_API_PYSIDE, 4),
-           pyqt5=(QT_API_PYQT5, 5), pyside2=(QT_API_PYSIDE2, 5))
-# ETS is a dict of env variable to (QT_API, QT_MAJOR_VERSION)
-# If the ETS QT_API environment variable is set, use it, but only
-# if the variable is of the same major QT version.  Note that
-# ETS requires the version 2 of PyQt4, which is not the platform
-# default for Python 2.x.
 
+QT_API_PYQT5 = "PyQt5"
+QT_API_PYSIDE2 = "PySide2"
+QT_API_PYQTv2 = "PyQt4v2"
+QT_API_PYSIDE = "PySide"
+QT_API_PYQT = "PyQt4"   # Use the old sip v1 API (Py3 defaults to v2).
 QT_API_ENV = os.environ.get('QT_API')
-
-if rcParams['backend'] == 'Qt5Agg':
-    QT_RC_MAJOR_VERSION = 5
-elif rcParams['backend'] == 'Qt4Agg':
-    QT_RC_MAJOR_VERSION = 4
-else:
-    # A different backend was specified, but we still got here because a Qt
-    # related file was imported. This is allowed, so lets try and guess
-    # what we should be using.
-    if "PyQt4" in sys.modules or "PySide" in sys.modules:
-        # PyQt4 or PySide is actually used.
-        QT_RC_MAJOR_VERSION = 4
-    else:
-        # This is a fallback: PyQt5
-        QT_RC_MAJOR_VERSION = 5
-
-QT_API = None
-
-# check if any binding is already imported, if so silently ignore the
-# rcparams/ENV settings and use what ever is already imported.
-if 'PySide' in sys.modules:
-    # user has imported PySide before importing mpl
-    QT_API = QT_API_PYSIDE
-
-if 'PySide2' in sys.modules:
-    # user has imported PySide before importing mpl
+# First, check if anything is already imported.
+if "PyQt5" in sys.modules:
+    QT_API = QT_API_PYQT5
+    dict.__setitem__(rcParams, "backend.qt5", QT_API)
+elif "PySide2" in sys.modules:
     QT_API = QT_API_PYSIDE2
+    dict.__setitem__(rcParams, "backend.qt5", QT_API)
+elif "PyQt4" in sys.modules:
+    QT_API = QT_API_PYQTv2
+    dict.__setitem__(rcParams, "backend.qt4", QT_API)
+elif "PySide" in sys.modules:
+    QT_API = QT_API_PYSIDE
+    dict.__setitem__(rcParams, "backend.qt4", QT_API)
+# Otherwise, check the QT_API environment variable (from Enthought).  This can
+# only override the binding, not the backend (in other words, we check that the
+# requested backend actually matches).
+elif rcParams["backend"] == "Qt5Agg":
+    if QT_API_ENV == "pyqt5":
+        dict.__setitem__(rcParams, "backend.qt5", QT_API_PYQT5)
+    elif QT_API_ENV == "pyside2":
+        dict.__setitem__(rcParams, "backend.qt5", QT_API_PYSIDE2)
+    QT_API = dict.__getitem__(rcParams, "backend.qt5")
```

Comment:
Is this correct even if `QT_API_ENV == 'foo'`?

Reply:
then we just respect whatever rcParams["backend.qt5"] originally was.